arch/arm/include/aarch32/cortex_m/tz_ns.h

```diff
@@ -0,0 +1,133 @@
+ *                  the functions have been called. This instruction must leave
+ *                  r0-r3 unmodified.
+ */
+#define __TZ_WRAP_FUNC_RAW(preface, name, postface, store_lr, load_lr) \
+	do { \
+		__asm(".global "#preface"; .type "#preface", %function"); \
+		__asm(".global "#name"; .type "#name", %function"); \
+		__asm(".global "#postface"; .type "#postface", %function"); \
+		\
+		__asm(store_lr); \
+		__asm("push {r0-r3}"); \
+		__asm("bl "#preface); \
+		__asm("pop {r0-r3}"); \
+		__asm("bl "#name); \
+		__asm("push {r0-r3}"); \
+		__asm("bl "#postface); \
+		__asm("pop {r0-r3}"); \
+		__asm(load_lr); \
+	} while (0)
+
+/**
+ * @brief Macro for "sandwiching" a function call (@p name) in two other calls
+ *
+ * This is for functions which do not pass arguments or return values via the
+ * stack. I.e. the arguments and return values must each fit within 4 words,
+ * after accounting for alignment.
+ *
+ * If nothing is passed on the stack, the stack can be used to store LR without
+ * messing with offsets.
+ *
+ *	int foo(char *arg); // Implemented elsewhere.
+ *	int __attribute__((naked)) foo_wrapped(char *arg)
+ *		{__TZ_WRAP_FUNC(bar, foo, baz)}
+ *
+ * is equivalent to
+ *
+ *	int foo(char *arg); // Implemented elsewhere.
+ *	int foo_wrapped(char *arg)
+ *	{
+ *		bar();
+ *		int res = foo(arg);
+ *		baz();
+ *		return res;
+ *	}
+ *
+ * @note __attribute__((naked)) is not mandatory, but without it, GCC gives a
+ *       warning for functions with a return value. It also saves a bit of space
+ *       since it removes a little code that is not necessary.
+ *
+ * See @ref __TZ_WRAP_FUNC_RAW for more information.
+ */
+#define __TZ_WRAP_FUNC(preface, name, postface) \
+	__TZ_WRAP_FUNC_RAW(preface, name, postface, "push {r4, lr}", \
+			"pop {r4, pc}")
+
+
+#ifdef CONFIG_ARM_FIRMWARE_USES_SECURE_ENTRY_FUNCS
+/**
+ * @brief Create a thread safe wrapper function for an non-secure entry function
+ *
+ * This locks the scheduler before calling the function by wrapping the NS entry
+ * function in @ref k_sched_lock / @ref k_sched_unlock, using
+ * @ref __TZ_WRAP_FUNC.
+ *
+ * In non-secure code:
+ *
+ *	int foo(char *arg); // Declaration of entry function.
+ *	TZ_THREAD_SAFE_ENTRY_FUNC(foo_safe, int, foo, char *arg)
+ *
+ * Usage in non-secure code:
+ *
+ *	int ret = foo_safe("my arg");
+ *
+ * If NS entry functions are called without such a wrapper, and a thread switch
+ * happens while execution is in the secure binary, the possibly app will crash
+ * upon returning to the non-secure binary.
+ *
+ * @param ret   The return type of the NS entry function.
+ * @param name  The desired name of the safe function. This assumes there is a
+ *              corresponding NS entry function called nsc_name.
+ * @param ...   The rest of the signature of the function. This must be the same
+ *              signature as the corresponding NS entry function.
+ */
+#define TZ_THREAD_SAFE_ENTRY_FUNC(name, ret, nsc_name, ...) \
```

Comment:
can you add "NONSECURE" as an in-fix in the name, here?

Reply:
Done